src/rules/ResourceSourceICUUnexplainedParams.js

```diff
@@ -0,0 +1,160 @@
+import Locale from "ilib-locale";
+import { IntlMessageFormat } from "intl-messageformat";
+import log4js from "log4js";
+
+// type imports
+/** @ignore @typedef {import("@formatjs/icu-messageformat-parser").MessageFormatElement} MessageFormatElement */
+/** @ignore @typedef {import("@formatjs/icu-messageformat-parser").ArgumentElement} ArgumentElement */
+/** @ignore @typedef {import("./ResourceSourceICUPluralCategories.js").Location} Location */
+
+const logger = log4js.getLogger("i18nlint.ResourceSourceICUUnexplainedParams");
+
+/**
+ * Check if replacement parameters mentioned in the English source string are
+ * mentioned in the description field as well. Translators can do a much better
+ * job of translating if they know what each replacement parameter represents.
+ */
+export class ResourceSourceICUUnexplainedParams extends ResourceRule {
+    /** @override */
+    name = "source-icu-unexplained-params";
+    /** @override */
+    description =
+        "Check if replacement parameters used in the source string are explained in the comments for translators.";
+    /** @override */
+    link =
+        "https://github.com/ilib-js/i18nlint/blob/main/docs/source-icu-unexplained-params.md";
+
+    /** @param {any} [opts] */
+    constructor(opts) {
+        super(opts);
+    }
+
+    /**
+     * @param {Object} params
+     * @param {string | undefined} params.source
+     * @param {Resource} params.resource
+     * @param {string} params.file
+     * @returns {Result[] | undefined}
+     * @override
+     */
+    matchString({ source, resource, file }) {
+        if (!source) return; // no source, no checks needed
+
+        const sourceLocale = new Locale(
+            resource.getSourceLocale() ?? this.sourceLocale
+        ).getSpec();
+
+        const resourceComment = resource.getComment();
+        if (!resourceComment) return; // no description provided, nothing to search
+
+        // attempt to parse the source string as IntlMessage and output any encountered errors
+        let ast;
+        try {
+            ast = new IntlMessageFormat(source, sourceLocale, undefined, {
+                captureLocation: true
+            }).getAst();
+        } catch (e) {
+            // this rule does not report on syntax errors in the source. Other rules do that,
+            // so we can ignore this exception
+            logger.debug(e);
+            return;
+        }
+
+        // find all replacement parameters
+        const /** @type {ArgumentElement[]} */ replacementParameters = [];
+        this.traverseIcuAst(ast, (element) => {
+            if (element.type === 1 /* argument */) {
+                replacementParameters.push(element);
+            }
+        });
+
+        // flag any replacement parameter whose label
+        // does not appear in the description of the resource
+        const missing = replacementParameters.filter(
+            (param) => !resourceComment.includes(param.value)
```

Comment:
I figured doing an exact match would decrease the amount of false passes - personally I feel it would make more sense to mention the replacement param exactly as it's used. I'll change it though, guess one could say it's better to output less false fails

Reply:
071e322